lighthouse-core/computed/metrics/lantern-first-contentful-paint.js

```diff
@@ -24,52 +26,109 @@ class LanternFirstContentfulPaint extends LanternMetric {
    */
-  static getOptimisticGraph(dependencyGraph, traceOfTab) {
-    const fcp = traceOfTab.timestamps.firstContentfulPaint;
-    const blockingScriptUrls = LanternMetric.getScriptUrls(dependencyGraph, node => {
-      return (
-        node.endTime <= fcp && node.hasRenderBlockingPriority() && node.initiatorType !== 'script'
-      );
+  static getBlockingCpuData(graph, fcpTs, networkFilter, cpuFilter) {
+    /** @type {Array<CPUNode>} */
+    const cpuNodes = [];
+    graph.traverse(node => {
+      if (node.type === BaseNode.TYPES.CPU && node.startTime <= fcpTs) {
+        cpuNodes.push(node);
+      }
     });
 
-    return dependencyGraph.cloneWithRelationships(node => {
-      if (node.endTime > fcp && !node.isMainDocument()) return false;
-      // Include EvaluateScript tasks for blocking scripts
-      if (node.type === BaseNode.TYPES.CPU) {
-        return node.isEvaluateScriptFor(blockingScriptUrls);
-      }
+    cpuNodes.sort((a, b) => a.startTime - b.startTime);
 
-      // Include non-script-initiated network requests with a render-blocking priority
-      return node.hasRenderBlockingPriority() && node.initiatorType !== 'script';
+    // A script is *possibly* render blocking if it finished loading before FCP
+    const possiblyRenderBlockingScriptUrls = LanternMetric.getScriptUrls(graph, node => {
+      return node.endTime <= fcpTs && networkFilter(node);
     });
+
+    // A script is *actually* render blocking if it finished loading before FCP *and* its
+    // EvaluateScript task finished before FCP as well.
+    /** @type {Set<string>} */
+    const actuallyRenderBlockingScriptUrls = new Set();
+    const blockingCpuNodeIds = new Set();
+    for (const url of possiblyRenderBlockingScriptUrls) {
+      for (const cpuNode of cpuNodes) {
+        if (cpuNode.isEvaluateScriptFor(new Set([url]))) {
+          actuallyRenderBlockingScriptUrls.add(url);
+          blockingCpuNodeIds.add(cpuNode);
+          break;
+        }
+      }
+    }
+
+    const firstLayout = cpuNodes.find(node => node.didPerformLayout());
+    if (firstLayout) blockingCpuNodeIds.add(firstLayout.id);
+    const firstPaint = cpuNodes.find(node => node.childEvents.some(e => e.name === 'Paint'));
+    if (firstPaint) blockingCpuNodeIds.add(firstPaint.id);
+    const firstParse = cpuNodes.find(node => node.childEvents.some(e => e.name === 'ParseHTML'));
+    if (firstParse) blockingCpuNodeIds.add(firstParse.id);
+
+    if (cpuFilter) cpuNodes.filter(cpuFilter).forEach(node => blockingCpuNodeIds.add(node.id));
+
+    return {possiblyRenderBlockingScriptUrls, actuallyRenderBlockingScriptUrls, blockingCpuNodeIds};
   }
 
   /**
    * @param {Node} dependencyGraph
-   * @param {LH.Artifacts.TraceOfTab} traceOfTab
+   * @param {number} fcpTs
+   * @param {function(NetworkNode):boolean} blockingScriptFilter
+   * @param {function(CPUNode):boolean=} cpuNodeFilter
```

Comment:
`cpuNodeFilter` does the opposite of what I expected...maybe incorporate `block`/`blocking` into its name as well?

Reply:
yeah this was a bad name, renamed both of them to `extraBlockingCpuNodesToIncludeFilter`